[Uptime] Query Overhaul + Pagination

andrewvc opened the pull request:
## Summary

Adds pagination to the Uptime app's monitor overview page. Also cleans up/removes a lot of unused code that laid the foundation for the states index approach we opted not to take.

This PR is somewhat complex due to our need to use composite aggregations with searchAfter style pagination.

# Understanding our new query approach

## How the schema works

At its simplest, we run a single heartbeat from a single location checking a single endpoint with a single IP.In this case, everytime we run a check we create a new document.

```
# Documents for a single monitor.id
# checked from one heartbeat
------ time --------------->
[D] [D] [D] [D] [D] [D] [D]
```

We can also check from multiple geo locations

```
# Documents for a single monitor.id
# checked from multiple heartbeats, on in a US-E datacenter, the other in US-W
Location | ---------- time ----------->
 US-E    ` [D] [D] [D] [D] [D] [D] [D]
 US-W    ` [D] [D] [D] [D] [D] [D] [D]
```


We can also check multiple ip addresses per check if there are multiple DNS entries. If the endpoint we're checking has 3 DNS entries we will create 3 documents per check. The final document has an extra `summary` field with two integer values: `summary.up` and `summary.down` summarizing how many documents were up vs down across all three documents in the check. All documents also contain a `monitor.check_group` UUID string field that has a shared value across all three documents.

```
# Documents for a single monitor.id checked from a single location checking 3 IP addresses total
# checked from one heartbeat.
# Documents with a summary are indicated with an [S]
Note the partially written group at the end
--------------- time --------------------->
Docs   | [D] [D] [S] [D] [D] [S] [D]
Groups | |--UUID1--| |--UUID2--| |-UUID3-
```

Putting it all together, we can have multiple geo locations with multiple IPs checked within each location.


```
# Documents for a single monitor.id checked from a single location checking 3 IP addresses total
# checked from one heartbeat.
# Documents with a summary are indicated with an [S]
Location | ------------------time-------------------------->
         `
US-W     ` [D] [D] [S] [D] [D] [S] [D]
         ` |--UUID1--| |--UUID2--| |-UUID3-
         `
US-E     ` [D] [D] [S]   [D] [D] [S] [D] [D] [S]
         ` |--UUIDA--|   |--UUIDB--| |--UUIDC--| 
```

## The Query

### Goals

Logically, what we want to do when searching is to:

1. Match the most recent complete check group from each location and return all documents from that check group.
1. For any specific fields (say `error.message` or `monitor.ip`) consider the monitor matched if any documents in any location have that value.
1. For status filtering consider the monitor `up` if all documents within the latest check groups for each location are up.

Additional properties we'd like this query to have:
1. Be paginatable
2. Be fast
3. Be in a consistent order (required for pagination)


### At a High Level

There are three phases, described in detail below, but the TL;DR is:

1. **queryPotentialMatches()** Find `monitor.id,mostrecent(check_group)` tuples that match the query and all filters except status. We eagerly fetch 500 results here knowing that we may discard some of these results in the next phase.
2. **refinePotentialMatches()** Perform an additional query that pulls in all the most recent check summaries from all locations for the monitor IDs from the last phase. We compare the matched check groups from the last phase to see if what matched previously was out of date. We can now apply the status filter as well. 
3. **enrich** At this point we already have the matching monitor IDs and their status, but we haven't brought in most of their fields. Here we re-query to get all docs from the matching monitor IDs most recent checks, then format them for return via API.

#### queryPotentialMatches()

This query returns via a composite agg all monitor IDs matching all query. It also returns the latest matching `monitor.check group` of all monitor IDs with the correct `monitor.status` value. Note that the returned `monitor.check_group` value will only be the most recent value matching the query terms. It may be an old check. 

So, the output of the check is all `monitor.id` values that have ever had a document that matched the query. However, these matches may be old.

**TODO: The PR as it stands doesn't work exactly this way, I added in an optimization for status filtering that actually is incorrect. We can exclude matches that are down if we're looking for things that are up, but nothing more than that**.

#### refinePotentialMatches()

In this phase we take the `monitor.id` values from the preceeding phase and query for all the latest check groups per geo-location for each via a terms query plus terms aggs, using a top hit to get the most recent (top hit size 1 is more efficient than a terms agg on a high cardinality string field sorted , I believe due to global ordinals). Then, in Javascript, we can further exclude monitors based on the `monitor.status` fiter. If the status filter is `up` we exclude any monitors that have any down checks using the `summary` fields. By using the `summary` fields we additionally ensure that we only include the most recent complete check groups, instead of partial ones.

We then combine the output of the preceding phase, kicking out results from the initial phase that didn't have their latest complete check groups match.

#### Enrichment

We already have the correct monitor IDs, but we don't have all the data from all matching documents inside the check groups. We only got the summary documents (and only returned the monitor.id and check group) in the preceding phase. In this straightforward phase we get all the documents for the latest check groups for each monitor and combine them into the final response. Additionally adding histogram data via an auxiliary query.


#### Pagination 

The query phases described above omit discussion of how pagination works. We don't want to retrieve all matching `monitor.ids`, if a user has 30k monitors that can be expensive. Our default page only shows 10 in this patch. By limiting results we increase speed. Additionally, by using "after" style pagination as `composite` aggregations use, each page load is as fast as the first, we don't have the performance issues traditional page number style pagination has as you go deep into the result set.

Internally the code has an Iterator style paginator object covering all phases except the final enrichment phase that has the following functions:

1. `next()` Gets the next monitor matching the query advances the internal cursor.
1. `peek()` Gets the next monitor matching the query, does not advance the internal cursor.
1. `current()` Gets the monitor at the current cursor
1. `paginationAfterCurrent()` Checks via peek to see if there is more data, and if so returns pagination data to be sent on the next request to get the next page. This powers the 'next' button.
1. `paginationBeforeCurrent()` Same as preceeding, but gets the pagination info for the page before the current item. **TODO add this method instead of the current hacky way with `reverse()`**

Internally we overfetch for the initial filter query to provide a buffer for the paginator; we don't want a query per `next()` invocation. That's currently set to 500 items at the moment, which is pretty snappy.


Fixes https://github.com/elastic/uptime/issues/63


## Alternative approaches considered

1. Using data frames: We rejected this approach for two reasons: 1.) It's one more thing that a user can accidentally break 2.) Worry over the cost of constantly re-indexing *every* document with low latency. We'd need to reindex everything once every 5-10s to prevent excessive lag. 
2. Kibana Background jobs and a high water mark. This would involve having a background job in Kibana that would compute the most recent documents for each monitor at a given time and then mark them with a high water mark using an update by query. This has the same issues as the data frames in terms of excessive doc writes and scalability. However, there's less to break in a sense in that there's no extra indices. OTOH, users have configure kibana to let it write to uptime indices.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

elasticmachine commented:
Pinging @elastic/uptime

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11498/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11629/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11640/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11806/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11841/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11847/)

justinkambic commented:
I'll document the contributions I make to this PR via comments since it's already quite far along. Initially when I loaded up the patch I noticed some unexpected behavior, there should only be six active monitors but my table had considerably more:

<img width="1879" alt="image" src="https://user-images.githubusercontent.com/18429259/62896240-9e0f9280-bd1e-11e9-97ee-2e7579e1ad41.png">

This is different than the behavior on master for me.

Additionally if we're going to have custom `forward`/`back` buttons we should probably place them on the lower-right of the block so it will conform with the location of the default pagination controls.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12268/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12269/)

justinkambic commented:
I've changed some code in https://github.com/elastic/kibana/pull/42933/commits/ae767ff3c6a4145d8ede72e13db4181d7fbfe942; I think the code was implemented the way it was to ensure that we allow users to click `Back` and `Forward` in the browser, but the existing URL params hook we maintain already causes React Router to push and pop with the history API.

Forward and back buttons are working for me with this implementation and I think the code is much cleaner. If there's some reason I'm missing for needing to do it the original way we can switch it back.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12321/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12483/)

justinkambic commented:
I noticed that when I applied filters like `monitor.status: up` to the page, I wasn't seeing the full set of `up` monitors. This took a significant amount of debugging to isolate, but it was caused by our choice of `after_key`. We were always choosing the `after_key` provided by Elasticsearch for our `check group` queries, no matter how many of the monitors we chose. Sometimes we would have more monitors than needed, or monitors would be filtered out. When we'd run subsequent `check group` queries, we'd start at the original `after_key`, thereby skipping monitors that weren't represented in a page yet.

To counteract this, I have added code in https://github.com/elastic/kibana/pull/42933/commits/d0d201cc3e055c0bb554a0f13fc022aaffd57c14 that persists the check group data. When a page is assembled, the last monitor on the page is read and the check group data for that monitor is used to create an `after_key`, ensuring that we will not skip monitors.

justinkambic commented:
I have noticed that when a filter is changed, the existing pagination parameters remain in the URL. In reality, these should be cleared any time the filtering parameter changes, because in reality we're assembling a new page each time. I added code in https://github.com/elastic/kibana/pull/42933/commits/3a2da31e11b67c2ced5870afe4f2982cbf6638d5 to handle this contingency.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12902/)

andrewvc commented on x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/elasticsearch_monitor_states_adapter.ts:
```diff
@@ -272,7 +274,15 @@ export class ElasticsearchMonitorStatesAdapter implements UMMonitorStatesAdapter
+               * it's important that the latest check group from the final monitor we use is what we return, or we will be way ahead in terms
+               * of check groups and end up skipping monitors on subsequent calls.
+               */
+              size: 500,
```
We should just make this 10000 which is the query max.

andrewvc commented on x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/elasticsearch_monitor_states_adapter.ts:
```diff
@@ -272,7 +274,15 @@ export class ElasticsearchMonitorStatesAdapter implements UMMonitorStatesAdapter
-              size,
+              /**
+               * TODO: extract this to a constant; we can't be looking for _only_
+               * ten monitors, because it's possible our check groups selection will represent more than ten.
```
Good catch

andrewvc commented on x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/elasticsearch_monitor_states_adapter.ts:
```diff
@@ -216,11 +217,12 @@ export class ElasticsearchMonitorStatesAdapter implements UMMonitorStatesAdapter
+        monitors = monitors.slice(0, STATES.LEGACY_STATES_QUERY_SIZE);
       }
-    } while (searchAfter !== null && monitors.length < STATES.LEGACY_STATES_QUERY_SIZE);
+    } while (!!searchAfter && monitors.length < STATES.LEGACY_STATES_QUERY_SIZE);
```
Do we actually need the double negation here? Shouldn't a plain `searchAfter` be the boolean equivalent?

justinkambic replied:
It should be, yes, I'm just in the habit of explicitly casting to a boolean. We don't need it. It's probably a leftover result of hacking more than anything else.

andrewvc commented on x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/elasticsearch_monitor_states_adapter.ts:
```diff
@@ -221,10 +227,28 @@ export class ElasticsearchMonitorStatesAdapter implements UMMonitorStatesAdapter
       monitors.push(...monitorBucketsToAdd);
       if (monitors.length > STATES.LEGACY_STATES_QUERY_SIZE) {
         monitors = monitors.slice(0, STATES.LEGACY_STATES_QUERY_SIZE);
+        isFinalPage = false;
```
Will this work if we have a page size divisible by the number of monitors, or will we have an empty final page as the result?

andrewvc replied:
Err trying this out that seems to be the case, I think we need to search for one extra item than we want then trim that item off.

justinkambic replied:
👍

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13219/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13625/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13698/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13702/)

andrewvc commented:
@justinkambic seems to work now, and with an updated query that's 100% accurate for the first time. All results are effectively queried over the most recent check group. Tested for perf with 10k monitors x2 locations. Still snappy!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13835/)

spalger commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13863/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/106/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1186/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1366/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1416/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1473/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1689/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1727/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2296/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2302/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2306/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2320/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2532/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2944/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3508/)

andrewvc commented on x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/search/enrich_monitor_groups.ts:
```diff
@@ -0,0 +1,386 @@
+    body: {
+      query: {
+        bool: {
+          filter: [{ terms: { 'monitor.check_group': checkGroups } }],
```
I attempted to only minimally alter this query and leave the obvious simplification here for a future PR since this one is big enough already. The only that's changed is that now we filter for exactly the docs we need (which of course makes most of the scripted metric redundant)

andrewvc replied:
Yes. Or, we can eliminate this stage complete and make this a per-monitor operation which I believe @shahzad31 is doing in a separate PR.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3545/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3616/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3768/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3784/)

shahzad31 commented:
@andrewvc i am unable to test this PR, because of this issue

![image](https://user-images.githubusercontent.com/3505601/65329017-0fb9e680-dbd1-11e9-923d-51669ece7ccf.png)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4088/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4109/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/271/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/308/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/316/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/425/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/600/)

andrewvc commented:
jenkins, retest this please

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/679/)

andrewvc commented on x-pack/legacy/plugins/uptime/public/components/functional/monitor_list/monitor_list.tsx:
```diff
@@ -60,37 +62,26 @@ export const MonitorListComponent = (props: Props) => {
+  const nextPagePagination = get<string>(data, 'monitorStates.nextPagePagination');
+  const prevPagePagination = get<string>(data, 'monitorStates.prevPagePagination');
+
+  const paginationLinkForSummary = (pagination: string | null, direction: string = 'next') => {
```
:+1:

andrewvc commented on x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/elasticsearch_monitor_states_adapter.ts:
```diff
@@ -4,607 +4,58 @@
-  }
+    statusFilter?: string
+  ): Promise<GetMonitorStatesResult> {
+    const size = 10;
```
Personally I'm not nuts about constants files just because in practice I rarely see the constants changed on any regular basis. They make the most sense IMHO when a number is shared between places in code in a way that must be coordinated.

andrewvc commented on x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/elasticsearch_monitor_states_adapter.ts:
```diff
@@ -4,607 +4,58 @@
-        size: pageSize,
-      },
+      pagination,
+      filterClause: filters && filters !== '' ? JSON.parse(filters) : null,
```
Well, this is just the `filters` clause. We just need to guard here so we don't `JSON.parse('')`

andrewvc commented on x-pack/legacy/plugins/uptime/public/components/functional/monitor_list/__tests__/monitor_list_pagination.test.tsx:
```diff
@@ -0,0 +1,128 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { shallowWithIntl } from 'test_utils/enzyme_helpers';
```
My thinking was that file was large enough already. If you feel strongly about it I'm OK to move it, it's not a giant deal t ome.

justinkambic replied:
It's ok to have them here.

andrewvc commented on x-pack/test/api_integration/apis/uptime/graphql/fixtures/ping_list_sort.json:
```diff
@@ -1,104 +1,7 @@
-        "observer": null
-      }
-    ]
+    "total": 0,
```
Hmmmm, this seems like an issue, I'll track this down.

andrewvc replied:
I believe this is fixed

andrewvc commented on x-pack/test/api_integration/apis/uptime/graphql/monitor_charts.js:
```diff
@@ -48,7 +47,8 @@ export default function ({ getService }) {
         .send({ ...getMonitorChartsQuery });
-      expect(data).to.eql(monitorChartsEmptySet);
+
+      expectFixtureEql(data, 'monitor_charts');
```
Nope, seems like an oversight on my part. Probably some copy and paste.

andrewvc commented on x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/search/fetch_page.ts:
```diff
@@ -0,0 +1,129 @@
+
+// Representation of a full page of results with pagination data for constructing next/prev links.
+export interface EnrichedPage {
+  items: MonitorSummary[];
```
I think it's a bit forward thinking in that @shahzad31 is working in a separate PR to make the monitor rows a separate API call. I think we are more likely to remove this functionality than add to it

andrewvc commented on x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/search/refine_potential_matches.ts:
```diff
@@ -0,0 +1,122 @@
+      const location = locBucket.key;
+      const topSource = locBucket.top.hits.hits[0]._source;
+      const checkGroup = topSource.monitor.check_group;
+
```
:+1:

andrewvc commented on x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/search/refine_potential_matches.ts:
```diff
@@ -0,0 +1,122 @@
+      size: 0,
+      query: {
+        bool: {
+          must: [
```
Great catch! Yes, no scoring

andrewvc commented on x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/search/refine_potential_matches.ts:
```diff
@@ -0,0 +1,122 @@
+          must: [
+            { terms: { 'monitor.id': monitorIds } },
+            // only match summary docs because we only want the latest *complete* check group.
+            { exists: { field: 'summary.down' } },
```
I had thought that wouldn't work because in lucene there are no objects, just k/v, but it somehow does!

andrewvc commented on x-pack/test/api_integration/apis/uptime/graphql/fixtures/monitor_charts.json:
```diff
@@ -1,49 +1,8 @@
 {
   "monitorChartsData": {
-    "locationDurationLines": [
```
No, def a bug.

andrewvc replied:
Is now fixed

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/1320/)

andrewvc commented on x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/search/enrich_monitor_groups.ts:
```diff
@@ -0,0 +1,386 @@
+  return sortedResItems;
+};
+
+const getHistogramForMonitors = async (
```
Should just be a straight copy/paste as much as possible.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/1348/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/1357/)

andrewvc commented:
@justinkambic I think I've addressed all the comments you left and have fixed the issues with empty test fixtures. The main source of problem there was date ranges that were too big (I just lazily selected 15 years), which resulted in no data since we drop single empty buckets. There were a couple of other tests that didn't match the new data right, and needed some of the test params tweaked.

Should be ready for a re-review.

I think to *really* test the new query even more extensively we have to stop relying on the `data.json.gz`, it's just too complex to reason about in testing. I think we need to synthetically generate minimal test cases within heartbeat.

My $0.02 is that we can perhaps merge this before doing that since this PR is so unwieldy, and tackle that as a follow-up ASAP.  I've opened an issue here to discuss this: https://github.com/elastic/kibana/issues/47001

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/1706/)

justinkambic commented:
It looks like most of my code concerns are resolved now. I'm doing some local smoke testing and things seem pretty consistent even with 13k+ monitors. I am noticing some slowness when filtering and paginating that many monitors, but I think it's acceptable for the time being.

justinkambic commented:
> My $0.02 is that we can perhaps merge this before doing that since this PR is so unwieldy, and tackle that as a follow-up ASAP

Agreed if we can get away from it that'd be ideal. We also should get https://github.com/elastic/kibana/pull/46269 merged before we tackle that.

shahzad31 commented:
i will spend some time again tomorrow morning to test it. But yes it seems like most of the suggestions reported are resolved.

andrewvc commented:
One thought here is that we could optimize the filtering of down monitors by not doing it in memory anymore. The tradeoff would be that users would need to know that filtering by a given set of search criteria and down would require that the down document meet that criteria.

I was initially against this, but the more I think about it the more I'm OK with it. WDYT?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/1815/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2082/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2399/)
* Commit: f4192d2826b57f68069238488f2941a69642bc44

justinkambic commented:
> The tradeoff would be that users would need to know that filtering by a given set of search criteria and down would require that the down document meet that criteria.

Does this mean that if the user filters by `name: monitor.name` and `down`, but the monitor is currently `up`, it will appear in the list but missing all its previous down checks?

andrewvc commented:
@justinkambic no, it would not appear at all because the refine phase would determine that the matched document from the potential phase is not the most recent one.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2528/)
* Commit: 6d18c71b7529349619cb7d3851ccfaec5a58790c

andrewvc commented:
jenkins, retest this please

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2558/)
* Commit: 6d18c71b7529349619cb7d3851ccfaec5a58790c

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2579/)
* Commit: a43d8a8f3741eddd7ebb06535ae8af3c00c4d88b

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2582/)
* Commit: ab9cc4919ac63812e8b6c1857d96c288543bc282

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2592/)
* Commit: 0a5c65ec462e05346665669561aca8fbd7ce02dc

andrewvc commented:
@justinkambic FYI I opened https://github.com/elastic/beats/pull/13929 to add index sorting to the beats side. One issue there potentially is that users will need to delete their templates or set heartbeat to overwrite their old one. Might be good to test for perf.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2632/)
* Commit: 953fb399d55fde71de9dd9e70e01acbc2c65f4bb

andrewvc commented:
@elasticmachine merge upstream

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2696/)
* Commit: 20f8be54d1d5601015e6c10effb213962bc47c2d

shahzad31 commented on x-pack/legacy/plugins/uptime/public/hooks/use_url_params.ts:
```diff
@@ -34,12 +34,26 @@ export const useUrlParams: UptimeUrlParamsHook = () => {
+          };
+        }, {})
+      ),
     });
```
Nice enhancement , it will clean URL up to some extent, i had already done this in my PR, but this looks even better :)

andrewvc commented:
LGTM to @justinkambic 's latest changes

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2920/)
* Commit: 8dfe20328d442aac6634e877ab297e48b69e7ec6

justinkambic commented:
Backported to:
`7.x`/`7.5.0` https://github.com/elastic/kibana/commit/b2b4212a2936121868ab45c6c244626653297aba
https://github.com/elastic/kibana/pull/48120
